File: c/library/zlib/stream_playback.py
import struct
import socket
import zlib

stream_header = 'h B h'

with open('7357.bak', 'rb') as f:
    header = f.read(2 + 1 + 2 + 16 + 2 + 16)
    ip_proto, is_ipv6 = struct.unpack('<HB', header[0:3])
    print 'ipproto:', ip_proto, 'is_ipv6:', is_ipv6
    if is_ipv6 == 1:
        pass
    else:
        server_port, server_ip = struct.unpack('<HI', header[3:9])
        client_port, client_ip = struct.unpack('<HI', header[21:27])
    print 'sport:', server_port, 'sip:', server_ip
    print 'cport:', client_port, 'cip:', client_ip
    print 'sip:', socket.inet_ntoa(header[5:9])
    print 'cip:', socket.inet_ntoa(header[23:27])

    dolog = f.read(4)
    dologid_count, = struct.unpack('<I', dolog)
    print 'dologid_count:', dologid_count
    dolog = f.read(dologid_count * 4)
    dologid = []
    for i in range(dologid_count):
        id, = struct.unpack('<I', dolog[i*4 : (i+1)*4])
        dologid.append(id)
    print 'dologid:', dologid

    data = f.read(4)
    total_out, = struct.unpack('<I', data)
    print 'total_out:', total_out
    data = f.read(total_out)
    dedata = zlib.decompress(data)
    size, = struct.unpack('<I', dedata[0:4])
    direction, = struct.unpack('<B', dedata[4:5])
    s2c_count, = struct.unpack('<I', dedata[5:9])
    c2s_count, = struct.unpack('<I', dedata[9:13])
    print 'size:', size, 'direction:', direction, \
            's2c_count:', s2c_count, 'c2s_count:', c2s_count
    print dedata[13:(13+size)]

File: c/library/zlib/stream_playback.c
#include <stdio.h>
#include <zlib.h>

#define STREAM_HEADER_LEN (sieof(Stream) - sizeof(char*))
#define DOLOGID_COUNT_MAX 100


static Stream sss;

static int InputStreamHeader(Stream* ss, FILE* input_file)
{
    char header[STREAM_HEADER_LEN];
    fread(header, 1, STREAM_HEADER_LEN, input_file);
    ss.ip_proto = (uint16_t*)(header); header += 2;
    ss.is_ipv6 = (uint8_t*)(header); header += 1;
    ss.sport = (uint16_t*)(header); header += sizeof(uint16_t);
    ss.cport = (uint16_t*)(header); header += sizeof(uint16_t);
    if (ss.is_ipv6) {
        ss.sip.h = (uint64_t*)(header); header += 8;
        ss.sip.l = (uint64_t*)(header); header += 8;
        ss.cip.h = (uint64_t*)(header); header += 8;
        ss.cip.l = (uint64_t*)(header); header += 8;
    }
    else {
        ss.sip = (uint32_t*)(header); header += 16;
        ss.cip = (uint32_t*)(header); header += 16;
    }
    return 0;
}

static int InputDologID(FILE* input_file)
{
    uint32_t count = 0;
    fread(&count, sizeof(uint32_t), 1, input_file);
    uint32_t dologid[DOLOGID_COUNT_MAX]; 
    fread(dologid, sizeof(uint32_t), count, input_file);
}

static int InputStreamPayload();

int PlayBack(FILE* input_file)
{
    Stream ss;
    while (true) {
        memset(&ss, 0, sizeof(ss)); 
        // header
        InputStreamHeader(&ss, input_file);
        // dolog
        InputDologID(input_file);
        while (true) {
            // stream
            PushStream() 
        }
    }

    return 0;
}

int main(int argc, char** argv)
{
    int err_code;
    if ((err_code = CreateEngine()) != ERROR_NORMAL) {
        printf("%s\n", GetErrorMsg(err_code));
        return -1;
    }

    FILE* input_file = fopen("1234.bak", "rb");
    PlayBack (input_file);

    DestroyEngine();

    return 0;
}
